lmon.1
.TH LMON 1
.SH NAME
lmon \- ncurses based linux performance viewer
.SH SYNOPSIS
.TP
.B lmon
[\fB\fIOPTION\fR]
.SH OPTIONS
.TP
.BR \-b
browse the history in the system lard database /var/lib/lard/lard.db (if present)
.TP
.BR \-f " " \fIdatabase\fR
browse the history in the lard \fIdatabase\fR
.TP
.BR \-g
regenerate the configuration file with defaults
.TP
.BR \-h
basic help.
.TP
.BR \-v
show lmon version.
.PP
without arguments, lmon operates as a realtime performance viewer.
.PP
.B lmon
likes terminals with 256 colors (such as TERM=xterm-256color) and
will adapt the screen layout to terminal resizing.
.SS "Exit status:"
.TP
0
if OK,
.TP
1
error.
.TP
2
more serious error.
.SH DESCRIPTION
.B lmon
shows kernel, CPU, memory, disk, filesystem, network and process activity
drawn from the /proc and /sys pseudo filesystems as well as udev.
.PP
.B lmon
either operates in realtime mode, taking snapshots from the live system,
or in browse mode (-f, -b), displaying snapshots stored in a lard database.
.PP
.B lmon
recognizes these key commands in realtime mode:
.TP
.BR \fBq
quits lmon (as does ^C).
.TP
.BR \fB+
increases sample interval by 1 second.
.TP
.BR \fB-
decreases sample interval by 1 second.
.PP
Note that changing the sample interval clears the CPU trail.
.PP
.B lmon
recognizes these key commands in browse mode:
.TP
.BR \fBq
quits lmon (as does ^C).
.TP
.BR \fBleftarrow " or " \fB,
move back
.TP
.BR \fBrightarrow " or " \fB.
move forward
.TP
.BR \fB+
zoom in
.TP
.BR \fB-
zoom out
.TP
.BR \fBh
one hour back
.TP
.BR \fBH
one hour forward
.TP
.BR \fBd
one day back
.TP
.BR \fBD
one day forward
.TP
.BR \fBw
one week back
.TP
.BR \fBW
one week forward
.TP
.BR \fBHOME
home key - start of history
.TP
.BR \fBEND
end key - end of history
.SH VIEWS
Some views are only visible, or partly visible, when the terminal screen is large enough.
.PP
Sizes expressed in bytes are given in base 1024, '1.0G' is a screen-space saving
abbreviation for 1.0GiB, one Gibibyte.
.SS "CPU"
CPU and schedulling details over the last sample interval, title shows
number of sockets/cores/logical CPU's. The character chart shows CPU mode
usage per logical CPU. The per-logical CPU chart is not shown when the number
of processors exceeds 32.
.PP
CPU utilization is given as the ratio CPU time / wallclock time, so
full utilization equals 1.
.TP
\fIuser\fR(\fIu\fR)
user mode CPU (CPU seconds/clock second).
.TP
\fInice\fR(\fIn\fR)
nice mode CPU (CPU seconds/clock second).
.TP
\fIsystem\fR(\fIs\fR)
system mode CPU (CPU seconds/clock second).
.TP
\fIiowait\fR(\fIw\fR)
iowait mode CPU (CPU seconds/clock second).
.TP
\fIirq\fR(\fIi\fR)
irq mode CPU (CPU seconds/clock second).
.TP
\fIsoftirq\fR(\fIo\fR)
softirq mode (CPU seconds/clock second).
.TP
\fItotal\fR
total of the above CPU modes.
.TP
\fIslice\fR
total CPU time divided by number of context switches.
.TP
\fIctxsw/s\fR
the number of context switches per second.
.TP
\fIforks/s\fR
the number of forks per second.
.TP
\fIloadavg\fR
5 and 10 minute load average (average number of processes on the run or block queue).
.TP
\fIrunq\fR
the number of processes on the run/block queue.
.SS "Memory"
Memory details, title shows the amount of real memory available
to the kernel.
.PP
Anonymous memory is memory not associated with file data.
.TP
.IR unused
real memory not in use.
.TP
.IR commitas
total virtual memory allocated.
.TP
.IR anon
anonymous virtual memory.
.TP
.IR file
real memory used as page cache.
.TP
.IR shmem
shared memory.
.TP
.IR slab
real memory used for the kernel slab.
.TP
.IR pagetbls
real memory used for virtual memory paging tables.
.TP
.IR dirty
memory used by dirty file pages.
.TP
.IR pgin/s
number of pages/s read into memory from persistent storage.
.TP
.IR pgout/s
number of pages/s written from memory to persistent storage.
.TP
.IR swpin/s
number of virtual memory pages/s read into memory from swap.
.TP
.IR swpout/s
number of virtual memory pages/s written from memory to swap.
.TP
\fIhp total
total number of hugepages.
.TP
\fIhp rsvd
number of reserved hugepages.
.TP
\fIhp free
number of free hugepages.
.TP
\fIthp anon
anonymous memory used by transparent hugepages.
.TP
.IR mlock
(m)locked memory.
.TP
.IR mapped
memory used by memory-mapped files.
.TP
\fIswp used
swap used.
.TP
\fIswp size
swap size.
.TP
.IR minflt/s
minor page faults per second.
.TP
.IR majflt/s
major page faults per second.
.TP
.IR alloc/s
memory allocations per second.
.TP
.IR free/s
memory freed per second.

.SS "System resources"
Miscellaneous system resources.
.TP
\fIfiles open
the number of allocated file handles on the system.
.TP
\fIfiles max
the maximum number of file handles that can be allocated on the system.
.TP
\fIinodes open
the number of allocated inode handles.
.TP
\fIinodes free
the number of free inode handles.
.TP
\fIprocesses
the number of processes.
.TP
\fIusers
the number of distinct users logged in.
.TP
\fIlogins
the number of user logins.
.TP
\fIfs growth/s
total filesystem growth per second.
.SS "CPU trail"
A time trail of CPU usage, aggregated over all CPUs. CPU modes are coded
by character and color. On the 'x-axis, a '+' marks 10 ticks.
.SS "Disk IO"
Disk IO statistics of top 'true' disks ordered by utilization. Derived block devices such as
LVM or MetaDisk are excluded. Not all disks may be shown due to lack
of terminal space, but the totals aggregate all disks nevertheless. The header shows
total attached storage size, average IOPS (r+w) and average bandwidth (r+w)
over the last sample interval.
.TP
\fI device
the device name.
.TP
\fI util
utilization as busy time / wallclock time.
.TP
\fI svct
average service time - not including queuing time.
.TP
\fI r/s
read operations per second.
.TP
\fI w/s
write operations per second.
.TP
\fI rb/s
bytes read per second.
.TP
\fI wb/s
bytes written per second.
.TP
\fI artm
average read time - includes queuing time.
.TP
\fI awtm
average write time - includes queuing time.
.TP
\fI rsz
average read size.
.TP
\fI wsz
average write size.
.TP
\fI qsz
number of IO's on the device queue at time of last sample.
.SS "Filesystem IO"
Shows statistics on block devices with mounted filesystems ordered by utilization. Same
columns as 'Disk IO'.
.SS "Network"
Shows network device statistics.
.TP
\fI device
the device name.
.TP
\fI rxb/s
bytes received per second.
.TP
\fI txb/s
bytes transmitted per second.
.TP
\fI rxpkt/s
packets received per second.
.TP
\fI txpkt/s
packets transmitted per second.
.TP
\fI rxsz
received packet average size.
.TP
\fI txsz
transmitted packet average size.
.TP
\fI rxerr/s
average receive error rate.
.TP
\fI txerr/s
average transmit error rate.
.SS "TCP server"
Shows TCP (v4 and v6) server connection statistics. Connections are counted and
grouped by (server address, server port, user running server process)
and sorted on number of connections.
.TP
\fI address
the address of the server.
.TP
\fI port
the port of the server.
.TP
\fI user
the user running the server process.
.TP
\fI #conn
the number of established tcp connections on this server.
.SS "TCP client"
Shows TCP (v4 and v6) client connection statistics. Client connections are counted and
grouped by (server address, server port, user running client process)
and sorted on number of connections.
.TP
\fI address
the address of the server.
.TP
\fI port
the port of the server.
.TP
\fI user
the user running the server process.
.TP
\fI #conn
the number of established tcp connections to the server.
.SS "Process"
Shows top processes ordered by \fItime\fR. In order to be 'seen', a process must exist
during at least two consecutive samples. Consequently, processes that are created
and destroyed within a sample interval are invisible and not aggregated.
The reported totals can therefore be lower than the system-wide reality shown in
the CPU and Memory views.
.PP
The Process view stops sampling when the sample time exceeds 120ms - the
sampling time scales linear with the number of processes on the system.
On an Intel core i7 this would disable the Process view output above
roughly 10000 processes. If the number of processes drops by 10% since disable, a new
sample is tried.
.TP
\fI pid
process id.
.TP
\fI pgrp
process group id.
.TP
\fI S
process status. In realtime mode, this is the status of the process
at the time of last sample. In browse mode, this is set to 'D' if
the pid ever had 'D' in the snapshot range.
.TP
\fI user
user owning the process.
.TP
\fI comm
command or 'process image'.
.TP
\fI time
total of utime, stime and iotime for the process.
.TP
\fI utime
user mode CPU time over the last interval divided by interval duration.
.TP
\fI stime
system mode CPU time over the last interval divided by interval duration.
.TP
\fI minflt
the minor faults per second caused by the process.
.TP
\fI majflt
the major faults per second caused by the process.
.TP
\fI rss
the process resident set size - real memory used by the process, some of
which may be shared.
.TP
\fI vsz
the process virtual memory size - virtual memory used by the process, some of
which may be shared.
.TP
\fI args
the process arguments.
.TP
\fI wchan
for blocked processes (status D), the kernel channel waited on. In
browse mode, this is set to the most frequent whcan for
pids having state D in the snapshot range at least once.
.SH CONFIG FILE
Configuration file is '.leanux-lmon' located in the first match among
 $XDG_CONFIG_HOME, HOME and getpwuid->pw_dir.
.SH BUGS
Report bugs, documentation errors and suggestions at https://github.com/jmspit/leanux/issues.
.SH AUTHOR
Jan-Marten Spit \<spitjm@xs4all.nl\>
.SH COPYRIGHT
copyright GPL v3, Jan-Marten Spit 2015-2022 <spitjm@xs4all.nl> https://jmspit.github.io/leanux
.PP
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.B lard(1), lblk(1), lrep(1), lsys(1)



tools/lmon/lmon.cpp
//========================================================================
//
// This file is part of the leanux toolkit.
//
// Copyright (C) 2015-2016 Jan-Marten Spit http://www.o-rho.com/leanux
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, distribute with modifications, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
// IN NO EVENT SHALL THE ABOVE COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR
// THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
// Except as contained in this notice, the name(s) of the above copyright
// holders shall not be used in advertising or otherwise to promote the
// sale, use or other dealings in this Software without prior written
// authorization.
//========================================================================

/**
 * @file
 * ncurses based realtime and historical performance viewer
 */
#include <iostream>
#include <iomanip>
#include <sstream>

#include "system.hpp"
#include "persist.hpp"
#include "util.hpp"
#include "lmon_curses.hpp"
#include "leanux-config.hpp"
#include "configfile.hpp"
#include "lard-config.hpp"

#include <signal.h>
#include <sys/time.h>
#include <algorithm>
#include <string.h>

namespace leanux {
  namespace tools {
    namespace lmon {

      /**
       * The global Screen.
       */
      Screen *screen = 0;

      /**
       * handle signals such a CTRL-C.
       */
      void  sig_handler(int sig) {
        signal(sig, SIG_IGN);
        if ( screen ) screen->Stop();
      }


      void printHelp() {
        std::cout << "lmon - Linux performance viewer." << std::endl;
        std::cout << "usage: lmon [OPTIONS]" << std::endl;
        std::cout << "  -b      : browse historic data from system lard database " << LARD_SYSDB_FILE << " (if present)" << std::endl;
        std::cout << "  -f file : browse historic data from a lard database file" << std::endl;
        std::cout << "  -g      : (re)write the configuration file with default values" << std::endl;
        std::cout << "  -h      : show this help" << std::endl;
        std::cout << "  -v      : show version" << std::endl;
        std::cout << std::endl;
        std::cout << "created by " << LEANUX_MAINTAINER_NAME << std::endl;
      }

      bool verifyDatabase( persist::Database &db ) {
        try {
          persist::Query qempty(db);
          qempty.prepare( "SELECT count(id) FROM snapshot" );
          if ( qempty.step() ) {
            long numsnaps = qempty.getLong(0);
            if ( numsnaps < 1 ) {
              std::cerr << "database stores 0 snapshots (lard running?)" << std::endl;
              return false;
            }
          } else {
            std::cerr << "invalid database" << std::endl;
            return false;
          }
        }
        catch ( const Oops& oops ) {
          std::cerr << "invalid database" << std::endl;
          return false;
        }
        return true;
      }

      int main( int argc, char* argv[] ) {

        #ifdef LEANUX_DEBUG
        util::Tracer::setTracer( "/tmp/lmon.trc" );
        #endif

        char* cur_term = getenv("TERM");
        std::stringstream s;
        s << "initial TERM value=" << cur_term;
        TRACE( s );

        signal( SIGINT, sig_handler );
        screen = 0;
        try {
          leanux::init();

          leanux::util::ConfigFile::declareParameter( "COLOR_BACKGROUND", "(10,10,12)", "Color for background (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_TEXT", "(90,120,170)", "Color for text (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_BOLD_TEXT", "(130,150,190)", "Color for bold text (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_LINE", "(50,50,73)", "Color for lines (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_RUNNING_PROC", "(1,234,109)", "Color for processes in R state (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_BLOCKED_PROC", "(239,40,40)", "Color for processes in D state (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_USER_CPU", "(200,234,109)", "Color for user mode CPU (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_NICE_CPU", "(70,234,70)", "Color for nice mode CPU (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_SYSTEM_CPU", "(237,89,29)", "Color for system mode CPU (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_WAIT_CPU", "(239,40,40)", "Color for wait mode CPU (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_IRQ_CPU", "(49,130,253)", "Color for irq mode CPU (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "COLOR_SOFTIRQ_CPU", "(117,164,234)", "Color for softirq mode CPU (only effective for 256color terminals)" );
          leanux::util::ConfigFile::declareParameter( "SAMPLE_INTERVAL", "4", "seconds between samples" );
          leanux::util::ConfigFile::declareParameter( "MAX_MOUNTPOINT_WIDTH", "22", "maximum characters used by a mountpoint" );
          leanux::util::ConfigFile::declareParameter( "IOVIEW_MIN_HEIGHT", "4", "minimum height (#rows) for disk and mount IO view" );
          leanux::util::ConfigFile::declareParameter( "IOVIEW_MAX_HEIGHT", "14", "maximum height (#rows) for disk and mount IO view" );
          leanux::util::ConfigFile::declareParameter( "NETVIEW_MIN_HEIGHT", "4", "minimum height (#rows) for network and TCP view" );

          leanux::util::ConfigFile::setConfig( "lmon", leanux::util::getUserConfigDir() + "/.leanux-lmon" );

          if ( argc > 1 ) {
            if ( strncmp( argv[1], "-h", 2 ) == 0 ) {
              printHelp();
            } else if ( strncmp( argv[1], "-v", 2 ) == 0 ) {
              std::cout <<  LEANUX_VERSION << std::endl;
            } else if ( strncmp( argv[1], "-g", 2 ) == 0 ) {
              leanux::util::ConfigFile::getConfig()->write(true);
            } else if ( strncmp( argv[1], "-f", 2 ) == 0 ) {
              if ( argc > 2 ) {
                if ( !util::fileReadAccess( argv[2] ) ) {
                  std::cerr << "database file '" << argv[2] << "' cannot be read" << std::endl;
                  return 1;
                } else {
                  persist::Database db( argv[2] );
                  if ( !verifyDatabase(db) ) return 1;
                  screen = new tools::lmon::Screen();
                  screen->runHistory( &db );
                  delete screen;
                }
              } else printHelp();
            } else if ( strncmp( argv[1], "-b", 2 ) == 0 ) {
              if ( !util::fileReadAccess( LARD_SYSDB_FILE ) ) {
                std::cerr << "database file '" << LARD_SYSDB_FILE << "' cannot be read" << std::endl;
                return 1;
              } else {
                persist::Database db( LARD_SYSDB_FILE );
                if ( !verifyDatabase(db) ) return 1;
                screen = new tools::lmon::Screen();
                screen->runHistory( &db );
                delete screen;
              }
            } else printHelp();
          } else {
            screen = new tools::lmon::Screen();
            screen->runRealtime();
            delete screen;
          }
        }
        catch ( Oops &oops ) {
          try {
            if ( screen ) {
              screen->Stop();
              delete screen;
              screen = 0;
            }
            std::cerr << "Oops:" << oops.getMessage() << std::endl;
            std::flush( std::cerr );
          }
          catch ( ... ) {
            std::cerr << "unhandled exception during handling of Oops: " << oops.getMessage() << std::endl;
            std::flush( std::cerr );
          }
          return 1;
        }
        catch ( ... ) {
          if ( screen ) {
            screen->Stop();
            delete screen;
            screen = 0;
          }
          std::cerr << "unhandled exception" << std::endl;
          std::flush( std::cerr );
          return 2;
        }

        return 0;
      }

    }; // namespace lmon
  }; // namespace tools
}; // namespace leanux

/**
 * lmon entry point.
 * @param argc number of arguments in argv.
 * @param argv arguments.
 * @return 0 on success, 1 on Oops, 2 on other errors.
 */
int main( int argc, char* argv[] ) {
  return leanux::tools::lmon::main( argc, argv );
}
